File: code/js/src/Domain/games/ship/Ship.ts
import {ShipType} from "./ShipType";
import {Orientation} from "./Orientation";
import {Coordinate} from "../Coordinate";

/**
 * A ship in the game.
 *
 * @property type the type of the ship
 * @property coordinate the coordinate of the ship
 * @property orientation the orientation of the ship
 */
export interface Ship {
    type: ShipType;
    coordinate: Coordinate;
    orientation: Orientation;
}

export class Ship {

    get coordinates(): Coordinate[] {
        return getCoordinates(this.type, this.coordinate, this.orientation);
    }

    constructor(type: ShipType, coordinate: Coordinate, orientation: Orientation) {
        this.type = type;
        this.coordinate = coordinate;
        this.orientation = orientation;
    }
}

/**
 * Returns the list of coordinates occupied by a ship.
 *
 * @param shipType the type of the ship
 * @param coordinate the coordinate of the ship
 * @param orientation the orientation of the ship
 *
 * @return the list of coordinates occupied by the ship
 */
export function getCoordinates(
    shipType: ShipType,
    coordinate: Coordinate,
    orientation: Orientation
): Coordinate[] {
    const coordinates: Coordinate[] = [];
    for (let i = 0; i < shipType.size; i++) {
        if (orientation === Orientation.HORIZONTAL)
            coordinates.push(new Coordinate(coordinate.col + i, coordinate.row));
        else if (orientation === Orientation.VERTICAL)
            coordinates.push(new Coordinate(coordinate.col, coordinate.row + i));
    }
    return coordinates;
}

/**
 * Checks if the given coordinate is valid for the given ship information.
 *
 * @param coordinate the coordinate
 * @param orientation the orientation of the ship
 * @param size the size of the ship
 * @param boardSize the size of the grid
 *
 * @return true if the coordinate is valid, false otherwise
 */
export function isValidShipCoordinate(
    coordinate: Coordinate,
    orientation: Orientation,
    size: number,
    boardSize: number
): boolean {
    const col = coordinate.col;
    const row = coordinate.row;
    const maxShipCol = (coordinate.col + size - 1)
    const maxShipRow = (coordinate.row + size - 1)

    return (
        orientation === Orientation.HORIZONTAL &&
        maxShipCol >= 1 && maxShipCol <= boardSize &&
        row >= 1 && row <= boardSize
    ) || (
        orientation === Orientation.VERTICAL &&
        maxShipRow >= 1 && maxShipRow <= boardSize &&
        col >= 1 && col <= boardSize
    );
}

File: code/jvm/src/main/kotlin/pt/isel/daw/battleships/domain/games/ship/Ship.kt
package pt.isel.daw.battleships.domain.games.ship

import pt.isel.daw.battleships.domain.games.Coordinate
import pt.isel.daw.battleships.domain.games.ship.Ship.Orientation.HORIZONTAL
import pt.isel.daw.battleships.domain.games.ship.Ship.Orientation.VERTICAL

/**
 * A Ship entity.
 * A ship can be undeployed or deployed.
 *
 * @property type the ship type
 * @property coordinate the coordinate of the ship
 * @property orientation the orientation of the ship
 * @property coordinates the ship's coordinates
 */
abstract class Ship(
    open val type: ShipType? = null, // Defaults needed for JPA
    open val coordinate: Coordinate? = null,
    open val orientation: Orientation? = null
) {

    val coordinates
        get() = getCoordinates(
            coordinate = coordinate ?: throw IllegalStateException("Ship has no coordinate"),
            size = type?.size ?: throw IllegalStateException("Ship has no type"),
            orientation = orientation ?: throw IllegalStateException("Ship has no orientation")
        )

    /**
     * Checks if this ship is overlapping with another ship.
     *
     * @param otherShip the other ship
     * @return true if the ships are overlapping, false otherwise
     */
    fun isOverlapping(otherShip: Ship): Boolean = otherShip.coordinates.any { it in coordinates }

    /**
     * The possible orientations of a ship.
     *
     * @property HORIZONTAL: the ship is horizontal
     * @property VERTICAL: the ship is vertical
     */
    enum class Orientation {
        HORIZONTAL,
        VERTICAL;
    }

    companion object {

        /**
         * Calculates the ship's coordinates.
         *
         * @property coordinate the coordinate of the ship
         * @property size the size of the ship
         * @property orientation the orientation of the ship
         *
         * @return the ship's coordinates
         */
        private fun getCoordinates(
            coordinate: Coordinate,
            size: Int,
            orientation: Orientation
        ): List<Coordinate> = (0 until size).map {
            when (orientation) {
                HORIZONTAL -> Coordinate(coordinate.col + it, coordinate.row)
                VERTICAL -> Coordinate(coordinate.col, coordinate.row + it)
            }
        }
    }
}
